_targets.R

```diff
@@ -0,0 +1,35 @@
+library(targets)
+source("code.R")
+tar_option_set(packages = c("tidyverse", "stringr", "sbtools", "whisker"))
+
+list(
+  # Get the data from ScienceBase
+  tar_target(
+    model_RMSEs_csv,
```

Comment:
@lindsayplatt The previous function I made that downloads data saves a file and stores the data into an R object. With this targets format, is it right that only the filename is stored? Is it possible to also store an R object that can be passed to other functions (e.g., a second name command that is the data object returned by the function)? I'm thinking I do not need the filename stored, and instead could work with only the R object.

Also, are you the course contact for these PRs?

Reply:
@jds485 For each target, you need to choose one of 3 types to track in the pipeline: an R object, a single file, or a directory of files. If you want to create a file but then also have an R object floating around your pipeline to use, I would suggest leaving this target as the one that creates the file (make sure that you have `return(out_filepath)` at the end of it) and then make a second target called `model_RMSEs` which calls `readr::read_csv()`. This way, you are tracking both the file and an object (but that is only if you want both, otherwise I would just plan to read in the CSV within your other functions).

> I'm thinking I do not need the filename stored

What is actually happening when you create file targets (by returning the filepath from the function & adding `format='file'`) is that `targets` is actually tracking the contents of the file, not just the filepath itself. So, if the _contents_ change, then it would rebuild. If you accidentally excluded `format='file'` in this target, then it would just be tracking the filepath and even if the file contents changed, it wouldn't properly kick off downstream steps. That is why including `format='file'` is critical for file targets!

Yes, I am the course contact for these PRs!

Reply:
Thanks! I think I understand now. Follow-up question: Is there an input/output efficiency loss by the "load data into R, save to file, read the file" structure? Or does targets somehow retain the initial downloaded data in memory (or the data structure, as in `vroom`)? I'm thinking for really large datasets it would be nice to only download or input the data once instead of twice.

Reply:
Underneath, `targets` is really storing all of the content as a file anyways (R objects get stored as `.rds` files in the `_targets/` directory), so either way data is getting written. I think we tend to lean more towards file targets but that is likely an artifact of how the previous pipeline tool worked. Since we often used file targets instead of objects, we did invest time into finding filetypes and functions that ease the read/write process and now lean towards faster/lighter filetypes (such as `.feather` files). You can read a wonderful breakdown of filetypes, sizes, and read/write speeds in [this blog post](https://waterdata.usgs.gov/blog/formats/) by Laura DeCicco. 

I think we could learn more about what the best approach is for `targets`, since it is newer to us. That could make for a great little side experiment - the cost of files vs objects for both size and speed.

Reply:
Loving this discussion thread. Popping in to note that `tar_target()` has the `format` argument that allows you to mess with the format with which `targets` saves an R object behind the scenes. Some are more efficient/fast/small than others depending on the format of your R object, so it can be worth it to think about which option you use for this argument. See https://www.rdocumentation.org/packages/targets/versions/0.4.2/topics/tar_target